### ISSUE TITLE
[Urgent] Shifting to an IaaS like Digital Ocean

### ISSUE THREAD

**Issue by ghost** [aneesh-neelam/VITacademics#32]
I am guessing you guys are using heroku, as far as my experience goes it gets really costly when you scale i suggest you guys shift to something like digital ocean its pricing is insanely cheap if you guys decide to i will be happy to help

**Comment by aneesh-neelam** [comment]
Digital Ocean is an Infrastructure-as-a-Service, a full-fledged hosting provider. Heroku is a Platform-as-a-Service. A PaaS would greatly reduce our efforts in maintenance and configuration. 

I do understand a PaaS is much more expensive, however scaling the Heroku app is unnecessary as we have tested and concluded that our biggest bottleneck is VIT's own server(s). Hence a single Heroku dyno will be enough for this. 

If VIT does manage to upgrade its infrastructure, we shall definitely move to a cheaper provider. Until then, I think it is unnecessary. Thank you for your suggestions.

**Comment by ghost** [comment]
I get that but if you use something like dokku:https://github.com/progrium/dokku you will have everything you need, i have used it before and it completely emulates heroku in digital ocean and it can be set up in minutes, and your shift to digital ocean in parallel there are more than enough guys in vit who can help you with that.

**Comment by aneesh-neelam** [comment]
Yes, I'm aware of that. I checked it out when we got free $100 platform credit for Digital Ocean with the [GitHub Student Developer Pack](https://education.github.com/pack).  But why is it necessary?

**Comment by sreeram-dev** [comment]
docker is also an alternative. 
@aashrairavooru  nodejs is a single process application which spawns light weight threads. PAAS are single process (with limited cpu/ram). the stack is made for each other. 

Regarding auto scaling, As number of cores increase. number of parallel threads increases. For a nodejs application you need more cpu cores and low ram instances.  Google Compute Engine (cheap and low latency) is awesome. 

Google Managed VMs is also an alternative. It gives you more flexibility and manages stuff like autoscaling.

**Comment by aneesh-neelam** [comment]
@aashrairavooru: Regarding Node.js on Heroku, please refer #21, the second post. We would need a load balancer with multiple application instances for scaling, exactly what the Heroku dynos and the Heroku router do.

**Comment by ghost** [comment]
You can scale using nginx for load balancing,port forwarding etc and nodejs has really improved in its cluster management as well and there are alternatives by strongloop for the same too and using google compute engine might not be a good idea it doesn't have its data centers in asia or asia-pacific digital ocean has one in singapore perfect for this case but if you have credits for the same its okay,consider rackspace too its pretty awesome

**Comment by ghost** [comment]
@sreeram-boyapati dokku is built upon docker and makes automated deployment really easy anyways we can do automated deployment ourselves using bare git repositories its damn easy https://www.digitalocean.com/community/tutorials/how-to-set-up-automatic-deployment-with-git-with-a-vps

**Comment by aneesh-neelam** [comment]
@aashrairavooru: Its not about it being easy. You can host a Node.js application using a number of providers, documented [here](https://github.com/joyent/node/wiki/node-hosting). I am aware of a number of them. 

We're hosting the Node.js app on Heroku and the MongoDB instance on MongoLab. We aren't paying a penny to them either, and neither of them require a credit card to be added for using the free plan. 

Like I mentioned earlier, we don't need to scale it to a paid plan because VIT's own server(s) cannot handle as much load as a single Heroku dyno running a Node.js app and a single sandbox MongoLab database. Hence, an IaaS is not necessary. 

If you do want to use cluster, check out [this file](https://github.com/aneesh-neelam/VITacademics/blob/master/bin/cluster) and modify [this file](https://github.com/aneesh-neelam/VITacademics/blob/master/bin/www)

**Comment by ghost** [comment]
I have faced crashes in the app before if it is not because of server overload then lets stick with it,but if you do find guys and even i will try to get some people on board will you guys be open to parallel migration and then a permanent shift to Digital ocean,as you said we do we have enough credits with us and i guess your project is about learning as well right then why not try something new which can be potentially benefitial for the app.

**Comment by aneesh-neelam** [comment]
@aashrairavooru: You faced a crash in [vitacademics-rel](http://vitacademics-rel.herokuapp.com/)? You will get a 503 error if the connection to VIT's website times out. 

IaaS providers like Digital Ocean usually do not have a free plan. We may have free credit for Digital Ocean due to the GitHub Education Pack, but that is limited and could be used for something more worthwhile or maybe something that actually requires it.

**Comment by aneesh-neelam** [comment]
We have to move to Digital Ocean, MongoLab's datalimits are being a massive pain in the ass.

**Comment by biocross** [comment]
is there any timeline for this?

**Comment by aneesh-neelam** [comment]
@karthikb351 already has an instance on Digital Ocean. Moving everything to that will take a couple of hours, that's it. 

And then the client apps will need to update the Base URL. We have that already too: https://vitacad.collegecode.in/

**Comment by biocross** [comment]
if you guys plan to do it within a day or two, i'll refrain from pushing an update for ios for bug fixes. let me know asap.

**Comment by aneesh-neelam** [comment]
Oh no, it's @karthikb351. Expect a week. :-P 
@biocross: Push your update, no rush.

**Comment by biocross** [comment]
Haha, alright.

**Comment by sreeram-dev** [comment]
Finally. I was waiting for this day

**Comment by aneesh-neelam** [comment]
Multiple MongoDB databases is making the entire backend very unstable. Atleast shift the database to DigitalOcean.

**Comment by aneesh-neelam** [comment]
The Heroku instance crashes a lot and stalls because it is unable to initiate and maintain 5 simultaneous MongoDB connection pools.
